### PR TITLE
adds a build task and updates readme

### PR DESCRIPTION
**Pull request by mynameisrufus**
This makes the output location set in stone and rather than document how to generate files we document it in the new `mix build`.

Some questions:

* If we are not checking in the generated code how does it get generated on `mix deps.get`?
* If we generate code on `mix deps.get` (I know it can be done) do we go to the internet to get all the xml files or check them into this repository?

### CODE REVIEW

File: config/config.exs [hunk 1]
```diff
@@ -1,5 +1,5 @@
 use Mix.Config
 
-config :mavlink, dialect: APM
+config :mavlink, dialect: MAVLink.Dialect.APM.Types
```
Comment (robinhilliard) [inline config/config.exs]: I'm not actually sure I'm supposed to include config.exs when it's packaged, these are for local testing in my working directory. My understanding is that when it's packaged the default env variables will be the ones in mix.exs application():

```
     env: [
        dialect: nil,
        system: 255,
        component: 250,
        connections: []
      ],
```
and users will override them with their own config.exs.

File: .gitignore [hunk 2]
```diff
@@ -16,11 +16,11 @@ erl_crash.dump
 # Also ignore archive artifacts (built via "mix archive.build").
 *.ez
 
-# Don't include generated MAVLink file
-MAVLink.ex
```
Comment (robinhilliard) [inline .gitignore:20]: Superseded by today's changes, not creating hardcoded MAVLink.ex anymore and don't need to ignore.

File: config/config.exs (hunk 1 above)
Comment (robinhilliard) [reply config/config.exs]: Also the dialect would be MAVLink.Dialect.APM without the Types

File: lib/mix/tasks/build_task.ex [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+  @doc """
+  
+  """
+  @shortdoc "Generate Elixir Module from MAVLink dialect XML"
+  @spec run([String.t]) :: :ok
+  @impl Mix.Task
+  
+  def run(_) do
+    MAVLink.Writer.create("test/input/ardupilotmega.xml", "lib/mavlink/dialect/apm.ex", "MAVLink.Dialect.APM")
```
Comment (robinhilliard) [inline lib/mix/tasks/build_task.ex]: I don't know about Tasks.Build, will have to read up on it.

Also more generally I'm not sure I can or should distribute the xml files with this library. Will have a look at pymavgen and what Tridge does. I don't want to create another source of truth for the mavlink definitions, they update every few days, and a mismatch of dialects is a safety issue I'd like users to take responsibility for. Perhaps mavlink message_definitions/v1.0 can be a git submodule or something?

Comment (mynameisrufus) [reply lib/mix/tasks/build_task.ex]: I'm not sure how git submodules work when you add a package as a dependency but will look into it. I have however seen ~https://github.com/bitwalker/timex~ https://github.com/lau/tzdata pull down a bunch of stuff when you add it as a dependency. We could do something similar and put the definitions somewhere, s3 bucket?

File: lib/mix/tasks/compile.mavlink.ex [hunk 4]
```diff
@@ -0,0 +1,22 @@
+defmodule Mix.Tasks.Compile.Mavlink do
+  use Mix.Task
+
+  def run(_args) do
+    HTTPoison.start()
+
```
Comment (mynameisrufus) [inline lib/mix/tasks/compile.mavlink.ex:6]: This is boilerplatey and I can impress my peers later with clever code if we want to take this approach of pulling things down in the compilation step.

**Comment by mynameisrufus** [comment]
Ok using a compile task to build the things, seems to work well. I did not remove the message definitions from the test dir as I don't know where to pull the `mini_mav....` stuff from.

File: lib/mix/tasks/build_task.ex (hunk 3 above)
Comment (robinhilliard) [reply lib/mix/tasks/build_task.ex]: Couldn't we pull them direct from mavlink github https://github.com/mavlink/mavlink/tree/master/message_definitions/v1.0

Comment (robinhilliard) [reply lib/mix/tasks/build_task.ex]: Although people can create their own definitions, wouldn't want to overwrite them. Someone might just want a robust comms library and not care about any of the drone stuff. Maybe overthinking...

Comment (mynameisrufus) [reply lib/mix/tasks/build_task.ex]: Yes we could pull directly, perhaps using tags.

I see so we would want something more dynamic, I will have a look.

Comment (robinhilliard) [reply lib/mix/tasks/build_task.ex]: Yes, more protobufs (or at least compiled from shared definition comms library, if that's what protobufs is) that happens to be used almost exclusively by remote vehicle community (at the moment)

**Pull request closed without merging**